Retry Action Scheduler jobs on timeout

nima-karimi opened the pull request:
### Changes proposed in this Pull Request:

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

Closes #808.

This PR adds a new method to the `ActionSchedulerJobMonitor` service that monitors the async jobs and reschedules them when they timeout.

This method uses PHP's `register_shutdown_function` to register a callback to run on PHP script failure due to a timeout error. There weren't any flags set in the error to specify it's a timeout so we need to search for the string `Maximum execution time` on the last thrown PHP error message (fetched by `error_get_last`) to check if it's a timeout error. This might make it a bit less reliable but I couldn't find any cases of PHP changing that error message.


A new filter is introduced for the failure timeframe and the previous failure rate threshold filter is renamed to be similar to our other filter names.

I've also reduced the job failure threshold to 3 (previously 5) and set the failure timeframe to 2 hours instead of 1. This means that if a job fails 3 times in the past two hours we won't retry it. This is to prevent retrying the timed-out jobs since they take longer to fail. 

The failure monitor is also refactored a bit so that it only watches for the exact job with the given arguments instead of generally watching the job's processing hook. Previously when a job failed multiple times we wouldn't process it anymore even when it was run with other arguments. Now, we also check the job's arguments and process them if the arguments have changed. I had some doubt here though because in some cases when a job is failing, it could fail for any given arguments (e.g. not configured correctly or API errors, etc.), but in other cases, a job's failure could be just for that specific set of args (e.g. non-existing or invalid product provided).


### Detailed test instructions:

1. Add `while (true) {}` in the beginning of the `UpdateAllProducts::process_items` method
2. Sync all products (async) via the connection test page
3. Go to `WooCommerce > Status > Scheduled Actions`
4. Wait for the job to time out and confirm that it gets rescheduled once it times out
5. Repeat steps 2 to 4 for three other times and confirm that after the last attempt it will not get rescheduled
6. Add `while (true) {}` in the beginning of the `UpdateProducts::process_items` method
7. Sync one product (async) via the connection test page
8. Repeat steps 3 to 5 this time for the Update Products job


<!--
Optional.
Enter a summary of all changes in this Pull Request, which will be added to the changelog if accepted.
Each line should start with `(Fix|Add|Tweak|Update) - `, for example:
> Fix - I took care of something that wasn't working.
> Add - I added something new that's pretty cool.
> Tweak - I made a small change.
> Update - I made big changes to something that wasn't broken.

Leave the "Changelog entry" header in place completely empty, without any summary if no changelog entry is needed.
If you remove the "Changelog entry" header, the title of Pull Request will be used as the changelog entry.
-->
### Changelog entry

> Tweak - Retry async jobs on timeout

nima-karimi committed:
Make the args nullable for `is_running`

nima-karimi committed:
Rename job failure rate filter

nima-karimi committed:
Change job failure rate threshold to 3

nima-karimi committed:
Add filter for job failure timeframe

nima-karimi committed:
Increase job failure timeframe to 2 hours

nima-karimi committed:
Refactor `validate_failure_rate`

This method now accepts `hook` and `args` argument and will only search for jobs with the specified arguments instead of just the job name.

nima-karimi committed:
Add method to monitor job timeout and retry.

This method uses PHP's `register_shutdown_function` to register a callback to run on PHP script failure due to timeout error. There wasn't any flags set in the error to specify it's a timeout so a `strpos` search is done on the PHP fatal errors to check if the timeout error has been thrown.

danielbitzer commented:
A couple of thoughts:

1. I think we may not want _all_ job types to automatically retry. For example, if a job was responsible for performing an action that we wouldn't want repeated e.g. sending emails or charging a user. At a glance it looks like in the case of this plugin all jobs could safely be re-run. However, it could be worth making it optional for failed tasks to be retried for the job type.
2. Wouldn't the timeout monitor need to be unhooked after the action is completed? If multiple actions are run in the same process and the last one times out then will all the actions that are being monitored be re-scheduled? You could consider using the `action_scheduler_unexpected_shutdown` hook instead or `shutdown` like in `\ActionScheduler_FatalErrorMonitor`.

nima-karimi committed:
Add a flag to optionally disable retry on timeout

nima-karimi committed:
Use ActionScheduler hook to monitor timeout

nima-karimi commented:
Thanks for the awesome suggestions, @danielbitzer 🥇 I didn't even realize that ActionScheduler already tracks the shutdown process and has a hook for unexpected errors. I used the `action_scheduler_unexpected_shutdown` in https://github.com/woocommerce/google-listings-and-ads/pull/1023/commits/471e76806cb563553bdc041ddd23452f93ac2026 and now it should be detached once the job is complete.

> I think we may not want all job types to automatically retry. 

As you mentioned, we currently don't have any jobs that are unsafe to retry and I didn't want to make things more complicated than we should. But just to make things a bit more flexible, I added a new class property to the abstract ActionScheduler job class in https://github.com/woocommerce/google-listings-and-ads/pull/1023/commits/efc72b3b25cb17dc7a216a5c20ed1d24ec95e13e which acts as a flag that allows the subclasses to disable the retry on timeout.

danielbitzer commented:
Looking good! 

Looking at this again I think we still need to unhook with `remove_action( 'action_scheduler_unexpected_shutdown' ...)` after the job is finished. My understanding is WordPress would still remember all the attached timeout monitors and attempt to run _all_ hooked functions rather than the most recent one...

nima-karimi committed:
Detach timeout monitor when the action ends.

nima-karimi committed:
Merge branch 'develop' into feature/retry-jobs-on-timeout

nima-karimi commented:
Thanks, @danielbitzer! I've refactored the code quite a bit in https://github.com/woocommerce/google-listings-and-ads/pull/1023/commits/496f0603bfc591c9c8e7f75206a6270de21e6c15. It now fires a separate action called `woocommerce_gla_action_scheduler_timeout` whenever an action scheduler job times out and then each job hooks into that action separately and registers its own `schedule` method with its set of arguments. Each job also detaches the reschedule action separately when the job is done. I believe this should resolve the issue. Let me know what you think.

danbitzer commented on src/Jobs/AbstractActionSchedulerJob.php:
```diff
@@ -82,6 +90,54 @@ public function handle_process_items_action( array $items ) {
+	 * @since x.x.x
+	 */
+	protected function detach_schedule_action_on_timeout( array $args ) {
+		remove_action(
```
This method of detaching a callable doesn't seem to work. I tried this and the callable didn't detach:

```php
function get_callable() {
	return function() {
		var_dump( 'test' );
	};
}

add_action( 'test_callable', get_callable() );

do_action( 'test_callable' );

remove_action( 'test_callable', get_callable() );

do_action( 'test_callable' );
```

danielbitzer reviewed (changes requested):
Looking good but I think there's an issue with how add/remove_action is being used...

nima-karimi committed:
Pass job args to timeout monitor

nima-karimi commented:
Hmm. That's weird because the closures returned by that function should have a matching object ID 🤔 Anyway, I used another workaround for removing the callback in https://github.com/woocommerce/google-listings-and-ads/pull/1023/commits/45ab6dfb5bec979461a443b523045584d74de783, which should work better. Instead of passing the job arguments to the callable, I passed them to the timeout monitor so that the `do_action` method passes them over to the callable. 

This resolves the issue with WP not being able to unhook closures from actions but it might cause another issue where if two jobs are running simultaneously and they timeout, we pass the arguments to their timeout callable without checking the job ID and it might cause conflicts. Not sure how often this could happen or how to resolve it yet. I kind of hit a roadblock here 😓 Let me know if you have a better idea for this. I might give this some more thought later.

nima-karimi commented:
One method would be to get the arguments from ActionScheduler instead of having each job pass its own. ActionScheduler passes an `action_id` to the shutdown action, we can use that to search for the set of arguments and pass those over to the timeout method. This gets a bit low level though and just seems like we're adding features to the ActionScheduler itself (come to think of it, it might not be a terrible idea).

danielbitzer commented:
> but it might cause another issue where if two jobs are running simultaneously and they timeout

It seems like this could happen since Action Scheduler allows concurrently running jobs. I'm not sure what the best way forward is either...

Could you get the `job_id` from the `action_hook` which could be gotten from the `action_id`?

> One method would be to get the arguments from ActionScheduler instead of having each job pass its own. ActionScheduler passes an `action_id` to the shutdown action, we can use that to search for the set of arguments and pass those over to the timeout method.

This makes sense. Right now I don't know how the `$args` are actually getting to the schedule function.

Also wouldn't this be necessary anyway to determine whether the timed-out action is `process_item` action vs a `create_batch` action?

nima-karimi committed:
Refactor the job timeout scheduler

Instead of relying on each job to provide their arguments, we fetch the job hook and arguments via ActionScheduler store using the ID that is passed to the `action_scheduler_unexpected_shutdown` action.

nima-karimi committed:
Check the job's hash before retry after timeout

A unique hash/checksum is generated for each job that is monitored for timeout and when a timeout occurs, we check that hash to make sure we are retrying the correct job (i.e. a job that is previously set to be monitored via `attach_timeout_monitor`).

nima-karimi commented:
I've refactored this quite a bit again. Now the job arguments and hook name are fetched using the `action_id` instead of having each job pass their own data. Moreover, to prevent any racing issues with simultaneous jobs being run in the same request I've added a class property to keep track of the jobs that are monitored. This means that only the monitored jobs will be retried even if they run parallel with other jobs in the same request. A checksum/hash is generated for each job using their arguments and hook name, which will be used to check if they have been attached to the monitor.

@danielbitzer ready for another review 🙏

danielbitzer reviewed (approved):
Looks good and works well!

I was wondering if it's worth logging when an action is retried due to a timeout to make it easier to debug on customer site's if something goes wrong. But not a big deal.

nima-karimi committed:
Log when retrying a job

nima-karimi committed:
Log if job failed too many times and not retrying on timeout

nima-karimi commented:
Thanks, Dan 🙇 I've added a log where we retry the job and also one where the job doesn't get rescheduled due to a high failure rate.
